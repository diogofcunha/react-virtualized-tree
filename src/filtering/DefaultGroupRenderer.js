import React from 'react';

const DefaultGroupRenderer = ({onChange, groups, selectedGroup}) => {
  return (
    <select
      className="tree-group"
      onChange={({target: {value}}) => {
        onChange(value);
      }}
      value={selectedGroup}
    >
      {Object.keys(groups).map(g => (
        <option key={g} value={g}>
          {groups[g].name}
        </option>
      ))}
    </select>
  );
};

export default DefaultGroupRenderer;
