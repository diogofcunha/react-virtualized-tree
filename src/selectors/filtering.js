const INITIAL_FILTERED_VALUE = {nodes: [], nodeParentMappings: {}};

export const filterNodes = (filter, nodes, parents = []) =>
  nodes.reduce((filtered, n) => {
    const {nodes: filteredChildren, nodeParentMappings: childrenNodeMappings} = n.children
      ? filterNodes(filter, n.children, [...parents, n.id])
      : INITIAL_FILTERED_VALUE;

    return !(filter(n) || filteredChildren.length)
      ? filtered
      : {
          nodes: [
            ...filtered.nodes,
            {
              ...n,
              children: filteredChildren,
            },
          ],
          nodeParentMappings: {
            ...filtered.nodeParentMappings,
            ...childrenNodeMappings,
            [n.id]: parents,
          },
        };
  }, INITIAL_FILTERED_VALUE);
