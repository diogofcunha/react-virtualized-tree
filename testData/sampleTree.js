export const COLLAPSED_CHILDREN = {
  id: 3,
  name: 'Leaf 3',
  state: {
    expanded: false,
    favorite: true,
    deletable: true,
  },
  children: [
    {
      id: 'c-3',
      name: 'Leaf 3 Child',
      state: {},
    },
  ],
};

export const EXPANDED_CHILDREN = {
  id: 2,
  name: 'Leaf 2',
  state: {
    expanded: true,
    deletable: true,
  },
  children: [
    COLLAPSED_CHILDREN,
    {
      id: 4,
      name: 'Leaf 4',
    },
  ],
};

export const EXPANDED_NODE_IN_ROOT = {
  id: 0,
  name: 'Leaf 1',
  state: {
    expanded: true,
  },
  children: [
    EXPANDED_CHILDREN,
    {
      id: 5,
      name: 'Leaf 5',
    },
  ],
};

export const COLLAPSED_NODE_IN_ROOT = {
  id: 1,
  name: 'Leaf 6',
  state: {
    expanded: false,
    deletable: true,
  },
  children: [
    {
      id: 6,
      name: 'Leaf 7',
      state: {
        expanded: false,
      },
      children: [
        {
          id: 7,
          name: 'Leaf 8',
        },
        {
          id: 8,
          name: 'Leaf 9',
        },
      ],
    },
    {
      id: 9,
      name: 'Leaf 10',
    },
  ],
};

export const DELETABLE_IN_ROOT = {
  id: 'z',
  name: 'Leaf z',
  state: {
    deletable: true,
    favorite: true,
  },
};

export const DELETABLE_CHILDREN = EXPANDED_CHILDREN;

export const Nodes = [
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
];
