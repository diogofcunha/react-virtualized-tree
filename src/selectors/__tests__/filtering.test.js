import {filterNodes} from '../filtering';
import {Nodes} from '../../../testData/sampleTree';

describe('filtering selectors', () => {
  const pairNodes = n => n.id % 2 === 0;

  it('should filter nodes based on injected filter', () => {
    expect(filterNodes(pairNodes, Nodes).nodes).toMatchSnapshot();
  });

  it('should create mappings matching filters', () => {
    const pairNodes = n => n.id % 2 === 0;

    expect(filterNodes(pairNodes, Nodes).nodeParentMappings).toMatchSnapshot();
  });
});
