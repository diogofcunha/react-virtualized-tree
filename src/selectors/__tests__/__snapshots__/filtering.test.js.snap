// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filtering selectors should create mappings matching filters 1`] = `
Object {
  "0": Array [],
  "1": Array [],
  "2": Array [
    0,
  ],
  "4": Array [
    0,
    2,
  ],
  "6": Array [
    1,
  ],
  "8": Array [
    1,
    6,
  ],
}
`;

exports[`filtering selectors should filter nodes based on injected filter 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "id": 4,
            "name": "Leaf 4",
          },
        ],
        "id": 2,
        "name": "Leaf 2",
        "state": Object {
          "deletable": true,
          "expanded": true,
        },
      },
    ],
    "id": 0,
    "name": "Leaf 1",
    "state": Object {
      "expanded": true,
    },
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "id": 8,
            "name": "Leaf 9",
          },
        ],
        "id": 6,
        "name": "Leaf 7",
        "state": Object {
          "expanded": false,
        },
      },
    ],
    "id": 1,
    "name": "Leaf 6",
    "state": Object {
      "deletable": true,
      "expanded": false,
    },
  },
]
`;
